README.md
# High Level Amazon S3 Client

## Installation

`npm install modern-s3`

Inspired by the now abandoned [s3](https://www.npmjs.com/package/s3)

## Features

 * Relies on the modern versions of AWS client SDK for retries. Does not implement its own retry logic
 * Includes logic to make multiple requests when there is a 1000 object limit.
 * Ability to set a limit on the maximum parallelization of S3 requests.
 * Ability to sync a dir to and from S3.
 * Promise based API with additional reporting if passed an event emitter
 * Supports files of any size (up to S3's maximum 5 TB object size limit).
 * Relies on the AWS SDK's own intelligent upload method to do multipart uploads.
 * Automatically provide Content-Type for uploads based on file extension.

## Synopsis

### Create a client

```js
import AWS from 'aws-sdk'
import * as s3 from 'modern-s3'

const awsClient = new AWS.S3({
  accessKeyId: "your s3 key",
  secretAccessKey: "your s3 secret",
  region: "your region",
  // See: http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Config.html#constructor-property
})

var client = s3.createClient({
  maxAsyncS3: 20,     // this is the default
  multipartUploadThreshold: 20971520, // this is the default (20 MB)
  multipartUploadSize: 15728640, // this is the default (15 MB)
  s3Client: awsClient
})
```

## API Documentation

### s3.createClient(options)

Creates an S3 client.

`options`:

 * `s3Client` - an instance of `AWS.S3`.
 * `maxAsyncS3` - maximum number of simultaneous requests this client will
   ever have open to S3. defaults to `20`.
 * `multipartUploadThreshold` - if a file is this many bytes or greater, it
   will be uploaded via a multipart request. Default is 20MB. Minimum is 5MB.
   Maximum is 5GB.
 * `multipartUploadSize` - when uploading via multipart, this is the part size.
   The minimum size is 5MB. The maximum size is 5GB. Default is 15MB. Note that
   S3 has a maximum of 10000 parts for a multipart upload, so if this value is
   too small, it will be ignored in favor of the minimum necessary value
   required to upload the file.

### s3.getPublicUrl(bucket, key, [bucketLocation])

 * `bucket` S3 bucket
 * `key` S3 key
 * `bucketLocation` string, one of these:
   - "" (default) - US Standard
   - "eu-west-1"
   - "us-west-1"
   - "us-west-2"
   - "ap-southeast-1"
   - "ap-southeast-2"
   - "ap-northeast-1"
   - "sa-east-1"

You can find out your bucket location programatically by using this API:
http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#getBucketLocation-property

returns a string which looks like this:

`https://s3.amazonaws.com/bucket/key`

or maybe this if you are not in US Standard:

`https://s3-eu-west-1.amazonaws.com/bucket/key`

### s3.getPublicUrlHttp(bucket, key)

 * `bucket` S3 Bucket
 * `key` S3 Key

Works for any region, and returns a string which looks like this:

`http://bucket.s3.amazonaws.com/key`


### client.deleteObjects(s3Params, ee?: EventEmitter)

See http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#deleteObjects-property

`s3Params` are the same.

The difference between using AWS SDK `deleteObjects` and this one:

 * Make multiple requests if the number of objects you want to delete is
   greater than 1000.

If passed an `EventEmitter` it emits the following events:

 * `'progress'` - emitted with object `{ progressAmount: number, progressTotal: number }`.
 * `'data' (data)` - emitted when a request completes.

Returns list of responses from AWS requests

### client.deleteDir(params)

Deletes specified directory and all its children in an S3 bucket. If objects within a directory
are more than 1000, it deletes them in multiple requests.

`params` is an object with following properties:
 * Bucket (required)
 * Prefix (required)
 * MFA (optional)

Returns list of responses from AWS Delete Object requests.

### client.listObjects(params, ee?: EventEmitter)

Lists objects recursively in the specified directory or path in S3. If there are more than 1000 requests, it
lists objects in multiple requests.

`params` is an object with following properties:
 * maxObjects: maximum number of objects to list (but might still return more than this number). Default is 10,0000
 * s3Params: See https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#listObjectsV2-property

If passed an EventEmitter, it emits the following events on it:

 * `'data'` - emitted when a request completes. Called with array of returned S3 objects (`AWS.S3.Object[]`)

Returns list of S3 Objects (`AWS.S3.Object[]`)

## Testing

`S3_KEY=<valid_s3_key> S3_SECRET=<valid_s3_secret> S3_BUCKET=<valid_s3_bucket> npm test`

Tests upload and download large amounts of data to and from S3.


CHANGELOG.md
### 0.1
First version. Has `listObjects`, `deleteDir` and `deleteObjects` methods only.


test/test.js
var s3 = require('../')
var MultipartETag = require('../lib/multipart_etag')
var path = require('path')
var ncp = require('ncp')
var Pend = require('pend')
var assert = require('assert')
var fs = require('fs')
var mkdirp = require('mkdirp')
var crypto = require('crypto')
var rimraf = require('rimraf')
var StreamSink = require('streamsink')
var tempDir = path.join(__dirname, 'tmp')
var tempManyFilesDir = path.join(__dirname, 'tmp', 'many-files-dir')
var localFile = path.join(tempDir, 'random.png')
var remoteRoot = 'node-s3-test/'
var remoteFile = remoteRoot + 'file.png'
var remoteFile2 = remoteRoot + 'file2.png'
var remoteFile3 = remoteRoot + 'file3.png'
var remoteDir = remoteRoot + 'dir1'
var remoteManyFilesDir = remoteRoot + 'many-files-dir'

var describe = global.describe
var it = global.it
var after = global.after
var before = global.before

var s3Bucket = process.env.S3_BUCKET

if (!s3Bucket || !process.env.S3_KEY || !process.env.S3_SECRET) {
  console.log('S3_BUCKET, S3_KEY, and S3_SECRET env vars needed to run tests')
  process.exit(1)
}

function createClient () {
  return s3.createClient({
    multipartUploadThreshold: 15 * 1024 * 1024,
    multipartUploadSize: 5 * 1024 * 1024,
    s3Options: {
      accessKeyId: process.env.S3_KEY,
      secretAccessKey: process.env.S3_SECRET,
      endpoint: process.env.S3_ENDPOINT
    }
  })
}

function createBigFile (file, size, cb) {
  mkdirp(path.dirname(file), function (err) {
    if (err) return cb(err)
    var md5sum = crypto.createHash('md5')
    var out = fs.createWriteStream(file)
    out.on('error', function (err) {
      cb(err)
    })
    out.on('close', function () {
      cb(null, md5sum.digest('hex'))
    })
    var str = 'abcdefghijklmnopqrstuvwxyz'
    var buf = ''
    for (var i = 0; i < size; ++i) {
      buf += str[i % str.length]
    }
    out.write(buf)
    md5sum.update(buf)
    out.end()
  })
}

function createFolderOfFiles (dir, numFiles, sizeOfFiles, cb) {
  for (var i = 0, j = numFiles; i < numFiles; i++) {
    createBigFile(path.join(dir, 'file' + i), sizeOfFiles, function () {
      j--
      if (j === 0) {
        cb()
      }
    })
  }
}

var file1Md5 = 'b1946ac92492d2347c6235b4d2611184'
describe('MultipartETag', function () {
  it('returns unmodified digest', function (done) {
    var inStream = fs.createReadStream(path.join(__dirname, 'dir1', 'file1'))
    var multipartETag = new MultipartETag()
    var bytes
    var progressEventCount = 0
    multipartETag.on('progress', function () {
      bytes = multipartETag.bytes
      progressEventCount += 1
    })
    multipartETag.on('end', function () {
      assert.ok(progressEventCount > 0)
      assert.strictEqual(bytes, 6)
      assert.strictEqual(multipartETag.digest.toString('hex'), file1Md5)
      assert.ok(multipartETag.anyMatch(file1Md5))
      assert.strictEqual(multipartETag.anyMatch(''), false)
      assert.strictEqual(multipartETag.anyMatch(null), false)
      done()
    })
    inStream.pipe(multipartETag)
    multipartETag.resume()
  })
})

describe('s3', function () {
  var hexdigest

  before(function (done) {
    var client = createClient()
    var s3Params = {
      Prefix: remoteRoot,
      Bucket: s3Bucket
    }
    var deleter = client.deleteDir(s3Params)
    deleter.on('end', function () {
      done()
    })
  })

  after(function (done) {
    rimraf(tempDir, done)
  })

  after(function () {
    fs.writeFileSync(path.join(__dirname, 'dir3', 'index.html'), '')
  })

  it('get public URL', function () {
    var httpsUrl = s3.getPublicUrl('mybucket', 'path/to/key')
    assert.strictEqual(httpsUrl, 'https://s3.amazonaws.com/mybucket/path/to/key')
    var httpUrl = s3.getPublicUrlHttp('mybucket', 'path/to/key')
    assert.strictEqual(httpUrl, 'http://mybucket.s3.amazonaws.com/path/to/key')
    // treat slashes literally
    httpsUrl = s3.getPublicUrl('marina-restaurant.at', 'uploads/about_restaurant_10.jpg', 'eu-west-1')
    assert.strictEqual(httpsUrl,
      'https://s3-eu-west-1.amazonaws.com/marina-restaurant.at/uploads/about_restaurant_10.jpg')
  })

  it('uploads', function (done) {
    createBigFile(localFile, 120 * 1024, function (err, _hexdigest) {
      if (err) return done(err)
      hexdigest = _hexdigest
      var client = createClient()
      var params = {
        localFile: localFile,
        s3Params: {
          Key: remoteFile,
          Bucket: s3Bucket
        }
      }
      var uploader = client.uploadFile(params)
      uploader.on('error', done)
      var progress = 0
      var progressEventCount = 0
      uploader.on('progress', function () {
        var amountDone = uploader.progressAmount
        var amountTotal = uploader.progressTotal
        var newProgress = amountDone / amountTotal
        progressEventCount += 1
        assert(newProgress >= progress, 'old progress: ' + progress + ', new progress: ' + newProgress)
        progress = newProgress
      })
      uploader.on('end', function (url) {
        assert.strictEqual(progress, 1)
        assert(progressEventCount >= 2, 'expected at least 2 progress events. got ' + progressEventCount)
        assert(url !== '', 'expected a url. got ' + url)
        done()
      })
    })
  })

  it('downloads', function (done) {
    doDownloadFileTest(done)
  })

  it('downloadBuffer', function (done) {
    var client = createClient()
    var downloader = client.downloadBuffer({ Key: remoteFile, Bucket: s3Bucket })
    downloader.on('error', done)
    var progress = 0
    var progressEventCount = 0
    var gotHttpHeaders = false
    downloader.on('progress', function () {
      var amountDone = downloader.progressAmount
      var amountTotal = downloader.progressTotal
      var newProgress = amountDone / amountTotal
      progressEventCount += 1
      assert(newProgress >= progress, 'old progress: ' + progress + ', new progress: ' + newProgress)
      progress = newProgress
    })
    downloader.on('httpHeaders', function (statusCode, headers, resp) {
      var contentType = headers['content-type']
      assert.strictEqual(contentType, 'image/png')
      gotHttpHeaders = true
    })
    downloader.on('end', function (buffer) {
      assert.strictEqual(progress, 1)
      assert(progressEventCount >= 3, 'expected at least 3 progress events. got ' + progressEventCount)
      var md5sum = crypto.createHash('md5')
      md5sum.update(buffer)
      assert.strictEqual(md5sum.digest('hex'), hexdigest)
      assert.ok(gotHttpHeaders)
      done()
    })
  })

  it('downloadStream', function (done) {
    var client = createClient()
    var downloadStream = client.downloadStream({ Key: remoteFile, Bucket: s3Bucket })
    downloadStream.on('error', done)
    var gotHttpHeaders = false
    downloadStream.on('httpHeaders', function (statusCode, headers, resp) {
      var contentType = headers['content-type']
      assert.strictEqual(contentType, 'image/png')
      gotHttpHeaders = true
    })
    var sink = new StreamSink()
    downloadStream.pipe(sink)
    sink.on('finish', function () {
      var md5sum = crypto.createHash('md5')
      md5sum.update(sink.toBuffer())
      assert.strictEqual(md5sum.digest('hex'), hexdigest)
      assert.ok(gotHttpHeaders)
      done()
    })
  })

  it('lists objects', function (done) {
    var params = {
      recursive: true,
      s3Params: {
        Bucket: s3Bucket,
        Prefix: remoteRoot
      }
    }
    var client = createClient()
    var finder = client.listObjects(params)
    var found = false
    finder.on('data', function (data) {
      assert.strictEqual(data.Contents.length, 1)
      found = true
    })
    finder.on('end', function () {
      assert.strictEqual(found, true)
      done()
    })
  })

  it('copies an object', function (done) {
    var s3Params = {
      Bucket: s3Bucket,
      CopySource: s3Bucket + '/' + remoteFile,
      Key: remoteFile2
    }
    var client = createClient()
    var copier = client.copyObject(s3Params)
    copier.on('end', function (data) {
      done()
    })
  })

  it('moves an object', function (done) {
    var s3Params = {
      Bucket: s3Bucket,
      CopySource: s3Bucket + '/' + remoteFile2,
      Key: remoteFile3
    }
    var client = createClient()
    var copier = client.moveObject(s3Params)
    copier.on('end', function (data) {
      done()
    })
  })

  it('deletes an object', function (done) {
    var client = createClient()
    var params = {
      Bucket: s3Bucket,
      Delete: {
        Objects: [
          {
            Key: remoteFile
          },
          {
            Key: remoteFile3
          }
        ]
      }
    }
    var deleter = client.deleteObjects(params)
    deleter.on('end', function () {
      done()
    })
  })

  it('uploads a folder', function (done) {
    var client = createClient()
    var params = {
      localDir: path.join(__dirname, 'dir1'),
      s3Params: {
        Prefix: remoteDir,
        Bucket: s3Bucket
      }
    }
    var uploader = client.uploadDir(params)
    uploader.on('end', function () {
      done()
    })
  })

  it('downloads a folder', function (done) {
    var client = createClient()
    var localDir = path.join(tempDir, 'dir-copy')
    var params = {
      localDir: localDir,
      s3Params: {
        Prefix: remoteDir,
        Bucket: s3Bucket
      }
    }
    var downloader = client.downloadDir(params)
    downloader.on('end', function () {
      assertFilesMd5([
        {
          path: path.join(localDir, 'file1'),
          md5: file1Md5
        },
        {
          path: path.join(localDir, 'file2'),
          md5: '6f0f1993fceae490cedfb1dee04985af'
        },
        {
          path: path.join(localDir, 'inner1/a'),
          md5: 'ebcb2061cab1d5c35241a79d27dce3af'
        },
        {
          path: path.join(localDir, 'inner2/b'),
          md5: 'c96b1cbe66f69b234cf361d8c1e5bbb9'
        }
      ], done)
    })
  })

  it('uploadDir with deleteRemoved', function (done) {
    var client = createClient()
    var params = {
      localDir: path.join(__dirname, 'dir2'),
      deleteRemoved: true,
      s3Params: {
        Prefix: remoteDir,
        Bucket: s3Bucket
      }
    }
    var uploader = client.uploadDir(params)
    uploader.on('end', function () {
      done()
    })
  })

  it('lists objects', function (done) {
    var params = {
      recursive: true,
      s3Params: {
        Bucket: s3Bucket,
        Prefix: remoteDir
      }
    }
    var client = createClient()
    var finder = client.listObjects(params)
    var found = false
    finder.on('data', function (data) {
      assert.strictEqual(data.Contents.length, 2)
      assert.strictEqual(data.CommonPrefixes.length, 0)
      found = true
    })
    finder.on('end', function () {
      assert.strictEqual(found, true)
      done()
    })
  })

  it('downloadDir with deleteRemoved', function (done) {
    var localDir = path.join(__dirname, 'dir1')
    var localTmpDir = path.join(tempDir, 'dir1')
    ncp(localDir, localTmpDir, function (err) {
      if (err) throw err

      var client = createClient()
      var params = {
        localDir: localTmpDir,
        deleteRemoved: true,
        s3Params: {
          Prefix: remoteDir,
          Bucket: s3Bucket
        }
      }
      var downloader = client.downloadDir(params)
      downloader.on('end', function () {
        assertFilesMd5([
          {
            path: path.join(localTmpDir, 'file1'),
            md5: 'b1946ac92492d2347c6235b4d2611184'
          },
          {
            path: path.join(localTmpDir, 'inner1/a'),
            md5: 'ebcb2061cab1d5c35241a79d27dce3af'
          }
        ], function (err) {
          if (err) throw err
          assert.strictEqual(fs.existsSync(path.join(localTmpDir, 'file2')), false)
          assert.strictEqual(fs.existsSync(path.join(localTmpDir, 'inner2/b')), false)
          assert.strictEqual(fs.existsSync(path.join(localTmpDir, 'inner2')), false)
          done()
        })
      })
    })
  })

  it('upload folder with delete removed handles updates correctly', function (done) {
    var client = createClient()
    var params = {
      localDir: path.join(__dirname, 'dir3'),
      deleteRemoved: true,
      s3Params: {
        Prefix: remoteDir,
        Bucket: s3Bucket
      }
    }
    var uploader = client.uploadDir(params)
    uploader.on('end', function () {
      // modify a file and upload again. Make sure the list is still intact.
      fs.writeFileSync(path.join(__dirname, 'dir3', 'index.html'), 'hi')
      var uploader = client.uploadDir(params)
      uploader.on('end', function () {
        var params = {
          recursive: true,
          s3Params: {
            Bucket: s3Bucket,
            Prefix: remoteDir
          }
        }
        var client = createClient()
        var finder = client.listObjects(params)
        var found = false
        finder.on('data', function (data) {
          assert.strictEqual(data.Contents.length, 2)
          assert.strictEqual(data.CommonPrefixes.length, 0)
          found = true
        })
        finder.on('end', function () {
          assert.strictEqual(found, true)
          done()
        })
      })
    })
  })

  it('uploads folder with lots of files', function (done) {
    createFolderOfFiles(tempManyFilesDir, 10, 100 * 1024, function () {
      var client = createClient()
      var params = {
        localDir: tempManyFilesDir,
        deleteRemoved: true,
        s3Params: {
          Prefix: remoteManyFilesDir,
          Bucket: s3Bucket
        }
      }
      var uploader = client.uploadDir(params)
      uploader.on('end', function () {
        // get a list of the remote files to ensure they all got created
        var params = {
          recursive: true,
          s3Params: {
            Bucket: s3Bucket,
            Prefix: remoteManyFilesDir
          }
        }
        var client = createClient()
        var finder = client.listObjects(params)
        var found = false
        finder.on('data', function (data) {
          assert.strictEqual(data.Contents.length, 10)
          found = true
        })
        finder.on('end', function () {
          assert.strictEqual(found, true)
          done()
        })
      })
    })
  })

  it('multipart upload', function (done) {
    createBigFile(localFile, 16 * 1024 * 1024, function (err, _hexdigest) {
      if (err) return done(err)
      hexdigest = _hexdigest
      var client = createClient()
      var params = {
        localFile: localFile,
        s3Params: {
          Key: remoteFile,
          Bucket: s3Bucket
        }
      }
      var uploader = client.uploadFile(params)
      uploader.on('error', done)
      var progress = 0
      var progressEventCount = 0
      uploader.on('progress', function () {
        var amountDone = uploader.progressAmount
        var amountTotal = uploader.progressTotal
        var newProgress = amountDone / amountTotal
        progressEventCount += 1
        assert(newProgress >= progress, 'old progress: ' + progress + ', new progress: ' + newProgress)
        progress = newProgress
      })
      uploader.on('end', function (data) {
        assert.strictEqual(progress, 1)
        assert(progressEventCount >= 2, 'expected at least 2 progress events. got ' + progressEventCount)
        assert.ok(data, 'expected data. got ' + data)
        done()
      })
    })
  })

  it('download file with multipart etag', function (done) {
    doDownloadFileTest(done)
  })

  it('deletes a folder', function (done) {
    var client = createClient()
    var s3Params = {
      Prefix: remoteRoot,
      Bucket: s3Bucket
    }
    var deleter = client.deleteDir(s3Params)
    deleter.on('end', function () {
      done()
    })
  })

  function doDownloadFileTest (done) {
    fs.unlink(localFile, function (err) {
      if (err) return done(err)
      var client = createClient()
      var params = {
        localFile: localFile,
        s3Params: {
          Key: remoteFile,
          Bucket: s3Bucket
        }
      }
      var downloader = client.downloadFile(params)
      downloader.on('error', done)
      var progress = 0
      var progressEventCount = 0
      var gotHttpHeaders = false
      downloader.on('progress', function () {
        var amountDone = downloader.progressAmount
        var amountTotal = downloader.progressTotal
        var newProgress = amountDone / amountTotal
        progressEventCount += 1
        assert(newProgress >= progress, 'old progress: ' + progress + ', new progress: ' + newProgress)
        progress = newProgress
      })
      downloader.on('httpHeaders', function (statusCode, headers, resp) {
        var contentType = headers['content-type']
        assert.strictEqual(contentType, 'image/png')
        gotHttpHeaders = true
      })
      downloader.on('end', function () {
        assert.strictEqual(progress, 1)
        assert(progressEventCount >= 3, 'expected at least 3 progress events. got ' + progressEventCount)
        var reader = fs.createReadStream(localFile)
        var md5sum = crypto.createHash('md5')
        reader.on('data', function (data) {
          md5sum.update(data)
        })
        reader.on('end', function () {
          assert.strictEqual(md5sum.digest('hex'), hexdigest)
          assert.ok(gotHttpHeaders)
          fs.unlink(localFile, done)
        })
      })
    })
  }
})

function assertFilesMd5 (list, cb) {
  var pend = new Pend()
  list.forEach(function (o) {
    pend.go(function (cb) {
      var inStream = fs.createReadStream(o.path)
      var hash = crypto.createHash('md5')
      inStream.pipe(hash)
      hash.on('data', function (digest) {
        var hexDigest = digest.toString('hex')
        assert.strictEqual(hexDigest, o.md5, o.path + ' md5 mismatch')
        cb()
      })
    })
  })
  pend.wait(cb)
}


lib/s3events.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var events_1 = require("events");
var S3EventEmitter = /** @class */ (function (_super) {
    __extends(S3EventEmitter, _super);
    function S3EventEmitter() {
        var _this = _super.call(this) || this;
        _this.progressAmount = 0;
        _this.progressTotal = 0;
        return _this;
    }
    return S3EventEmitter;
}(events_1.EventEmitter));
exports.default = S3EventEmitter;


lib/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var url_1 = __importDefault(require("url"));
var events_1 = require("events");
exports.MAX_PUTOBJECT_SIZE = 5 * 1024 * 1024 * 1024;
exports.MAX_DELETE_COUNT = 1000;
exports.MAX_MULTIPART_COUNT = 10000;
exports.MIN_MULTIPART_SIZE = 5 * 1024 * 1024;
function sleep(delay) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, new Promise(function (resolve) { return setTimeout(resolve, delay); })];
        });
    });
}
exports.createClient = function (options) {
    return new Client(options);
};
var Client = /** @class */ (function () {
    function Client(options) {
        options = options || {};
        this.s3 = options.s3Client;
        if (!options.s3Client) {
            throw new Error('You need to pass in your own initialized s3Client');
        }
        this.maxAsyncS3 = options.maxAsyncS3 || 20;
        this.s3RetryCount = options.s3RetryCount || 3;
        this.s3RetryDelay = options.s3RetryDelay || 1000;
        this.multipartUploadThreshold = options.multipartUploadThreshold || (20 * 1024 * 1024);
        this.multipartUploadSize = options.multipartUploadSize || (15 * 1024 * 1024);
        this.multipartDownloadThreshold = options.multipartDownloadThreshold || (20 * 1024 * 1024);
        this.multipartDownloadSize = options.multipartDownloadSize || (15 * 1024 * 1024);
        if (this.multipartUploadThreshold < exports.MIN_MULTIPART_SIZE) {
            throw new Error('Minimum multipartUploadThreshold is 5MB.');
        }
        if (this.multipartUploadThreshold > exports.MAX_PUTOBJECT_SIZE) {
            throw new Error('Maximum multipartUploadThreshold is 5GB.');
        }
        if (this.multipartUploadSize < exports.MIN_MULTIPART_SIZE) {
            throw new Error('Minimum multipartUploadSize is 5MB.');
        }
        if (this.multipartUploadSize > exports.MAX_PUTOBJECT_SIZE) {
            throw new Error('Maximum multipartUploadSize is 5GB.');
        }
    }
    Client.prototype.deleteObjects = function (s3Params, ee) {
        return __awaiter(this, void 0, void 0, function () {
            var params, slices, progressTotal;
            var _this = this;
            return __generator(this, function (_a) {
                params = {
                    Bucket: s3Params.Bucket,
                    Delete: __assign({}, s3Params.Delete),
                    MFA: s3Params.MFA
                };
                slices = chunkArray(params.Delete.Objects, exports.MAX_DELETE_COUNT);
                progressTotal = 0;
                return [2 /*return*/, Promise.all(slices.map(function (slice) { return __awaiter(_this, void 0, void 0, function () {
                        var data, progressAmount;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0:
                                    params.Delete.Objects = slice;
                                    return [4 /*yield*/, this.s3.deleteObjects(params).promise()];
                                case 1:
                                    data = _a.sent();
                                    if (ee && data.Deleted) {
                                        progressAmount = data.Deleted.length;
                                        progressTotal += progressAmount;
                                        ee.emit('progress', { progressAmount: progressAmount, progressTotal: progressTotal });
                                    }
                                    ee && ee.emit('data', data);
                                    return [2 /*return*/, data];
                            }
                        });
                    }); }))];
            });
        });
    };
    Client.prototype.deleteDir = function (s3Params) {
        return __awaiter(this, void 0, void 0, function () {
            var bucket, listObjectsParams, ee, listObjectsPromise, deleteObjectPromises;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        bucket = s3Params.Bucket;
                        listObjectsParams = {
                            s3Params: {
                                Bucket: bucket,
                                Prefix: s3Params.Prefix
                            }
                        };
                        ee = new events_1.EventEmitter();
                        listObjectsPromise = this.listObjects(listObjectsParams, ee);
                        deleteObjectPromises = [];
                        ee.on('data', function (objects) {
                            var deleteParams = {
                                Bucket: s3Params.Bucket,
                                Delete: {
                                    Objects: objects.map(keyOnly),
                                    Quiet: true
                                },
                                MFA: s3Params.MFA
                            };
                            deleteObjectPromises.push(_this.s3.deleteObjects(deleteParams).promise());
                        });
                        return [4 /*yield*/, listObjectsPromise];
                    case 1:
                        _a.sent();
                        return [2 /*return*/, Promise.all(deleteObjectPromises)];
                }
            });
        });
    };
    Client.prototype.listObjects = function (params, ee) {
        return __awaiter(this, void 0, void 0, function () {
            var s3Details, MAX_KEYS, maxObjectsToList, s3ObjectsList, isPending, continuationToken, listData, listObjects;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        s3Details = __assign({}, params.s3Params);
                        MAX_KEYS = 1000;
                        maxObjectsToList = params.maxObjects || 10 * 1000;
                        s3ObjectsList = [];
                        isPending = true;
                        continuationToken = s3Details.ContinuationToken;
                        _a.label = 1;
                    case 1:
                        if (!(isPending && s3ObjectsList.length < maxObjectsToList)) return [3 /*break*/, 3];
                        return [4 /*yield*/, this.s3.listObjectsV2(__assign(__assign({}, s3Details), { ContinuationToken: continuationToken, MaxKeys: s3Details.MaxKeys || MAX_KEYS })).promise()];
                    case 2:
                        listData = _a.sent();
                        if (!listData.Contents) {
                            throw new Error('List Contents should always be defined');
                        }
                        isPending = !!listData.IsTruncated;
                        continuationToken = listData.NextContinuationToken;
                        listObjects = listData.Contents;
                        if (ee) {
                            ee.emit('data', listObjects);
                        }
                        s3ObjectsList = __spreadArrays(s3ObjectsList, listObjects);
                        return [3 /*break*/, 1];
                    case 3: return [2 /*return*/, s3ObjectsList];
                }
            });
        });
    };
    return Client;
}());
function chunkArray(array, maxLength) {
    var slices = [array];
    while (slices[slices.length - 1].length > maxLength) {
        slices.push(slices[slices.length - 1].splice(maxLength));
    }
    return slices;
}
function encodeSpecialCharacters(filename) {
    // Note: these characters are valid in URIs, but S3 does not like them for
    // some reason.
    return encodeURI(filename).replace(/[!'()* ]/g, function (char) {
        return '%' + char.charCodeAt(0).toString(16);
    });
}
function getPublicUrl(bucket, key, bucketLocation, endpoint) {
    var nonStandardBucketLocation = (bucketLocation && bucketLocation !== 'us-east-1');
    var hostnamePrefix = nonStandardBucketLocation ? ('s3-' + bucketLocation) : 's3';
    var parts = {
        protocol: 'https:',
        hostname: hostnamePrefix + '.' + (endpoint || 'amazonaws.com'),
        pathname: '/' + bucket + '/' + encodeSpecialCharacters(key)
    };
    return url_1.default.format(parts);
}
exports.getPublicUrl = getPublicUrl;
function getPublicUrlHttp(bucket, key, endpoint) {
    var parts = {
        protocol: 'http:',
        hostname: bucket + '.' + (endpoint || 's3.amazonaws.com'),
        pathname: '/' + encodeSpecialCharacters(key)
    };
    return url_1.default.format(parts);
}
exports.getPublicUrlHttp = getPublicUrlHttp;
function keyOnly(item) {
    return {
        Key: item.Key,
        VersionId: item.VersionId
    };
}
